File: dart3d/android/src/main/kotlin/com/jasonholtdigital/dart3d/FlutterAssets.kt
package com.jasonholtdigital.dart3d

import android.content.Context

/**
 * W25 fix-2: one resolver for every wire asset `ref`. DartNative —
 * like Flutter — nests the pubspec `assets/` tree under
 * `assets/flutter_assets/` inside the APK, so a verbatim
 * `context.assets.open(ref)` misses packaged assets. Try the nested
 * path first, then the bare ref (non-Flutter-packaged asset trees
 * still resolve); a ref already carrying the prefix skips the
 * double-prefix.
 */
object FlutterAssets {
    fun readBytes(context: Context, ref: String): ByteArray? {
        if (ref.isEmpty()) return null
        val paths = if (ref.startsWith("flutter_assets/")) {
            arrayOf(ref)
        } else {
            arrayOf("flutter_assets/$ref", ref)
        }
        for (path in paths) {
            try {
                return context.assets.open(path).use { it.readBytes() }
            } catch (e: Exception) {
                // Miss — try the next shape.
            }
        }
        return null
    }
}

File: dart3d/ios/Classes/FlutterAssets.swift
import Foundation

/// W25 fix-2: one resolver for every wire asset `ref`. DartNative —
/// like Flutter — nests the pubspec `assets/` tree at
/// `Frameworks/App.framework/flutter_assets/<ref>` inside the app
/// bundle, so a bare `Bundle.main.url(forResource:)` misses packaged
/// assets. Try the nested path first, then the bundle root (non-
/// Flutter-packaged resources still resolve); a ref already carrying
/// the prefix skips the double-prefix.
enum FlutterAssets {
    static func url(forResource ref: String) -> URL? {
        guard !ref.isEmpty else { return nil }
        if let dir = Bundle.main.privateFrameworksPath {
            let nested = ref.hasPrefix("flutter_assets/") ? ref
                : "flutter_assets/" + ref
            let url = URL(fileURLWithPath: dir)
                .appendingPathComponent("App.framework")
                .appendingPathComponent(nested)
            if FileManager.default.fileExists(atPath: url.path) {
                return url
            }
        }
        return Bundle.main.url(forResource: ref, withExtension: nil)
    }

    static func data(forResource ref: String) -> Data? {
        url(forResource: ref).flatMap { try? Data(contentsOf: $0) }
    }
}
